src/stable_hasher.rs

```diff
@@ -8,48 +8,113 @@ use std::hash::Hasher;
 #[cfg(test)]
 mod tests;
 
-/// Trait for retrieving the result of the stable hashing operation.
-pub trait StableHasherResult: Sized {
-    fn finish(hasher: StableHasher) -> Self;
-}
-
+/// A Stable Hasher adapted for cross-platform independent hash.
+///
 /// When hashing something that ends up affecting properties like symbol names,
 /// we want these symbol names to be calculated independently of other factors
 /// like what architecture you're compiling *from*.
 ///
 /// To that end we always convert integers to little-endian format before
 /// hashing and the architecture dependent `isize` and `usize` types are
 /// extended to 64 bits if needed.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+/// use std::hash::Hasher;
+///
+/// struct Hash128((u64, u64));
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hasher: StableHasher) -> Hash128 {
+///         Hash128(hasher.finalize())
+///     }
+/// }
+///
+/// let mut hasher = StableHasher::new();
+/// hasher.write_usize(0xFA);
+///
+/// let hash: Hash128 = hasher.finish();
+/// ```
+#[must_use]
 pub struct StableHasher {
     state: SipHasher128,
 }
 
-impl fmt::Debug for StableHasher {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:?}", self.state)
-    }
+/// Trait for retrieving the result of the stable hashing operation.
+///
+/// # Example
+///
+/// ```
+/// use rustc_stable_hash::{StableHasher, StableHasherResult};
+///
+/// struct Hash128((u64, u64));
+///
+/// impl StableHasherResult for Hash128 {
+///     fn finish(hasher: StableHasher) -> Hash128 {
+///         Hash128(hasher.finalize())
+///     }
+/// }
+/// ```
+pub trait StableHasherResult: Sized {
+    /// Retriving the [`finalize`]d state of the [`StableHasher`] and construct
+    /// an [`Self`] containing the hash.
+    ///
+    /// [`finalize`]: StableHasher::finalize
+    fn finish(hasher: StableHasher) -> Self;
 }
 
 impl StableHasher {
+    /// Creates a new [`StableHasher`].
+    ///
+    /// To be used with the [`Hasher`] implementation and [`StableHasher::finish`].
     #[inline]
+    #[must_use]
     pub fn new() -> Self {
         StableHasher {
             state: SipHasher128::new_with_keys(0, 0),
         }
     }
 
+    /// Returns the typed-hash value for the values written.
+    ///
+    /// The resulting typed-hash value is constructed from an
+    /// [`StableHasherResult`] implemenation.
+    ///
+    /// To be used in-place of [`Hasher::finish`].
     #[inline]
+    #[must_use]
     pub fn finish<W: StableHasherResult>(self) -> W {
         W::finish(self)
     }
 
+    /// Returns the hash value for the values written so far.
+    ///
+    /// This method returns a `(u64, u64)` pair for greater precision.
+    ///
+    /// See [`StableHasher::finish`] for an idomatic use of this function.
```

Comment:
this point doesn't make sense to me. if you use `finish` you don't just this function...

Reply:
It's used indirectly. Calling `StableHasher::finish` will call `StableHasherResult::finish` which should call `StableHasher::finalize`.

Reply:
I looked all 3 implementation of `StableHasherResult` in `rustc` and they all immediately call `StableHasher::finalize` in their `StableHasherResult::finish` implementation, we could therefore conceivably remove `StableHasher::finalize` and directly pass the `(u64, u64)` tuple in `StableHasherResult::finish`; which would make this less confusing and simpler to understand. Thoughs?

Reply:
Could I still do `let hash: SomethingImplemetingStableHasherResult = hasher.finish();`?

Reply:
> Could I still do `let hash: SomethingImplemetingStableHasherResult = hasher.finish();`?

Yes. The only thing that will change if we remove `StableHasher::finalize` is the signature of `StableHasherResult::finish` which will go from `fn finish(hasher: StableHasher) -> Self;` to `fn finish(hash: (u64, u64)) -> Self;`

Reply:
Is this what you mean?

```rust
pub trait StableHasherResult: Sized {
    fn from(hash: (u64, u64)) -> Self;
}
```

Reply:
> Is this what you mean?

Yes

Reply:
Whoops, I didn't see your reply 🙂
That seems like a fine change to me.

Reply:
> It's used indirectly. 

yes, but "idiomatic use" implied that you actually call the function directly.

the above change to the result trait to accept a `[u64; 2]` makes sense to me.

Reply:
> the above change to the result trait to accept a `[u64; 2]` makes sense to me.

I posted https://github.com/rust-lang/rustc-stable-hash/pull/4 to address this.